firedrake/plot.py

```diff
@@ -292,6 +294,321 @@ def quiver(function, **kwargs):
+            next_chunk = next_chunk[1:, :]
+            next_index = self._index_of_first_bad_point(next_chunk)
+
+            self._enter_distance_to_chunk(chunk[:index, :])
+            chunks.append(chunk[:index, :])
+
+            chunk, reason, index = next_chunk, next_reason, next_index
+
+        if index != 0:
+            self._enter_distance_to_chunk(chunk[:index])
+            chunks.append(chunk[:index])
+
+        backward_streamline = np.concatenate(chunks, axis=0)
+
+        streamline = np.vstack((backward_streamline[:0:-1], forward_streamline))
+        self.streamlines.append(streamline)
+
+
+def streamplot(function, resolution=None, min_length=None, max_time=None,
+               start_width=0.5, end_width=1.5, tolerance=3e-3, loc_tolerance=1e-10,
+               seed=None, **kwargs):
+    r"""Create a streamline plot of a vector field
+
+    Similar to matplotlib :func:`streamplot <matplotlib.pyplot.streamplot>`
+
+    :arg function: the Firedrake :class:`~.Function` to plot
+    :arg resolution: minimum spacing between streamlines (defaults to domain size / 20)
+    :arg min_length: minimum length of a streamline (defaults to 4x resolution)
+    :arg max_time: maximum time to integrate a streamline
+    :arg start_width: line width at beginning of streamline
+    :arg end_width: line width at end of streamline, to convey direction
+    :arg tolerance: dimensionless tolerance for adaptive ODE integration
+    :arg loc_tolerance: point location tolerance for :meth:`~firedrake.functions.Function.at`
+    :kwarg kwargs: same as for matplotlib :class:`~matplotlib.collections.LineCollection`
+    """
+    if function.ufl_shape != (2,):
+        raise ValueError("Streamplot only defined for 2D vector fields!")
+
+    axes = kwargs.pop("axes", None)
+    if axes is None:
+        figure = plt.figure()
+        axes = figure.add_subplot(111)
+
+    mesh = function.ufl_domain()
+    if resolution is None:
+        coords = mesh.coordinates.dat.data_ro
+        resolution = (coords.max(axis=0) - coords.min(axis=0)).max() / 20
+
+    if min_length is None:
+        min_length = 4 * resolution
+
+    if max_time is None:
+        area = assemble(Constant(1) * dx(mesh))
+        average_speed = np.sqrt(assemble(inner(function, function) * dx) / area)
+        max_time = 50 * min_length / average_speed
+
+    streamplotter = Streamplotter(function, resolution, min_length, max_time,
+                                  tolerance, loc_tolerance)
+
+    # TODO: better way of seeding start points
+    start_points = []
+    shape = streamplotter._grid.shape
+    for i in range(shape[0] - 1):
+        for j in range(shape[1] - 1):
+            start_points.append(streamplotter._grid_point((i, j)))
+
+    # Randomly shuffle the start points
+    import randomgen
+    generator = randomgen.MT19937(seed).generator
+    start_points = generator.permutation(np.array(start_points))
+
+    for x in start_points:
+        streamplotter.add_streamline(x)
```

Comment:
Given that you add all the points, what is the rationale for permuting them? Is it that streamlines are merged in the order they're seen?

Minor, just do:
```python
for x in generator.permutation(...):
    streamplotter.add_streamline(x)
```

Reply:
The algorithm tries every grid point to initialize a streamline, but many of them will get "covered" by an earlier streamline and thus not used at all. I've found that randomizing the insertion order usually starts the streamlines far apart at first, which gives you longer streamlines. The Right Thing To Do is to use farthest point seeding, but that requires spatial data structures we don't have if you want it to run in less than O(n^2). Randomized approaches happen to work pretty well and could be improved on using [Poisson disk sampling](https://www.jasondavies.com/poisson-disc/).

Reply:
So we have some r-trees lying around, but I don't know what spatial data structures you need.